### PR TITLE
Feature/local 3w map

### PR DESCRIPTION
**Pull request by frozenhelium**

### CODE REVIEW

File: src/root/views/GlobalThreeW/Map/index.tsx [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  ongoing_projects: number;
+  target_total: number;
+  society_name: string;
+  name: string;
+  projects_per_sector: {
+    primary_sector: number;
+    primary_sector_display: string;
+    count: number;
+  }[];
+}
```
Comment (tnagorra) [inline src/root/views/GlobalThreeW/Map/index.tsx:69]: Looks like this is also defined on GlobalThreeW

File: src/root/components/ExpandableContainer/index.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import Container, {
+  Props as ContainerProps,
+} from '#components/Container';
+
+export type Props = Omit<ContainerProps, 'headerElementRef'> & {
+  initiallyExpanded?: boolean;
+};
+
+function ExpandibleContainer(props: Props) {
```
Comment (tnagorra) [inline src/root/components/ExpandableContainer/index.tsx:20]:
```suggestion
function ExpandableContainer(props: Props) {
```

File: src/root/components/spinner.js [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import React from 'react';
+import { _cs } from '@togglecorp/fujs';
 
-function Spinner () {
+function Spinner (p) {
```
Comment (tnagorra) [inline src/root/components/spinner.js:4]:
```suggestion
function Spinner (props) {
```

File: src/root/utils/common.ts [hunk 4]
```diff
@@ -38,6 +38,20 @@ export function transformObjectItems<K extends string, T, R>(obj: Record<K, T>,
   }), {} as Record<K, R>);
 }
 
+export function denormalizeList<A, B, C>(
+  list: A[],
+  secondaryListSelector: (li: A) => B[],
+  transformFn: (li: A, sli: B) => C,
```
Comment (tnagorra) [inline src/root/utils/common.ts:44]: We should add proper type definitions

File: src/root/views/Country/ThreeW/InCountryProjects/Map/index.tsx [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+          },
+        };
+      }).filter(isDefined),
+    }),
+    [districtList, districtGroupedProjects],
+  );
+
+  const selectedDistrictProjectDetail = React.useMemo(
+    () => {
+      if (!clickedPointProperties || !clickedPointProperties.feature?.id) {
```
Comment (tnagorra) [inline src/root/views/Country/ThreeW/InCountryProjects/Map/index.tsx:144]:
```suggestion
      if (!clickedPointProperties?.feature?.id) {
```

File: src/root/views/Country/ThreeW/InCountryProjects/Map/index.tsx [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  );
+
+  const handlePointClose = React.useCallback(
+    () => {
+      setClickedPointProperties(undefined);
+    },
+    [setClickedPointProperties],
+  );
+
+  const maxScaleValue = Math.max(projectList?.length ?? 0, 2);
```
Comment (tnagorra) [inline src/root/views/Country/ThreeW/InCountryProjects/Map/index.tsx:177]: We may need to add comment why we did Math.max with value 2

Comment (tnagorra) [reply src/root/views/Country/ThreeW/InCountryProjects/Map/index.tsx:177]: Maybe we should move this Math.max inside the helper utility where the value `2` makes sense

File: src/root/views/Country/ThreeW/InCountryProjects/index.tsx [hunk 7]
```diff
@@ -64,16 +84,33 @@ function InCountryProjects(props: Props) {
     query: {
       limit: 500,
       country: country?.iso,
+      ...filters,
+      project_districts: filters.project_districts.length > 0 ?
+        filters.project_districts : undefined,
```
Comment (tnagorra) [inline src/root/views/Country/ThreeW/InCountryProjects/index.tsx:89]:
```suggestion
      project_districts: filters.project_districts.length > 0
          ? filters.project_districts
          : undefined,
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/root/views/GlobalThreeW/Map/index.tsx [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    clickedPointProperties,
+    setClickedPointProperties,
+  ] = React.useState<ClickedPoint| undefined>();
+
+  const allCountries = useReduxState('allCountries');
+
+  const countries = allCountries?.data?.results;
+
+  const maxProjectCount = React.useMemo(() => (
+    max(projectList, d => d.ongoing_projects)
```
Comment (AdityaKhatri) [inline src/root/views/GlobalThreeW/Map/index.tsx:149]: We might need to filter out undefined

File: src/root/views/Country/ThreeW/NSProjects/index.tsx [hunk 9]
```diff
@@ -28,29 +36,44 @@ import {
...
+
 interface Props {
   country: Country | undefined;
   className?: string;
   projectsUpdatedOn?: number | undefined;
 }
 
 function NSProjects(props: Props) {
+  const user = useReduxState('me');
+  const isLoggedIn = !!user.data.id;
```
Comment (AdityaKhatri) [inline src/root/views/Country/ThreeW/NSProjects/index.tsx:62]:
```suggestion
  const isLoggedIn = !!user?.data?.id;
```

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

**Pull request merged**